src/Typin/Typin/Experimental/Design.cs

```diff
@@ -0,0 +1,174 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Typin.Experimental
+{
+    // Command Specification
+
+    /// <summary>
+    /// This interface represents a type of command where its <see cref="Descriptor"/> describes 
+    /// the class of command (with summary, options, arguments, examples, etc.).
+    /// 
+    /// This command type could represent just a generic description for multiple other sub-commands,
+    /// and not be able to run itself. However it could add some description and examples for the entire
+    /// collection of sub-commands, and add some common arguments and options accesible for all sub-commands
+    /// 
+    /// A <see cref="ICommandProvider"/> returns a collection if this type to indicate available 
+    /// commands to the Typin runtime.
+    /// </summary>
+    public interface ICommandType
```

Comment:
In intent, yes, they represent the concept of a command's schema, so this could be `ICommandSchema` for that matter.

But I wouldn't want to associate a `CommandSchema` with a `Type` always. I have not digged deep enough in your code so you should know better than me. The reason there is a Type associated to a CommandSchema, is to `GetRequiredService(schema.Type)` to be able to get an instance of `ICommand`.

That is why I proposed two separate interfaces `ICommandType` and `IExecutableCommandType` (maybe renamed as `ICommandSchema` and `IExecutableCommandSchema`, or `IAbstractCommandSchema` and `ICommandSchema`). 

The former as an `Abstract command schema`, with a descriptor, but no other task to perform other than showing its help, when invoked from the command line. Take as an example `dotnet watch`, it shows it's help, and have a couple of sub-commands, and parameters and options usable from all their sub-commands

The second kind of schema, are for commands that CAN perform some task. In which case they have the posibility to `CreateRunner`. So in your world, your `ICommand` is my `ICommandRunner`. Maybe `ICommandTask` or others are more appropriate, but this is a first draft. I wanted to do it this way because I know I have some discrepancies and wanted the opportunity to present my side so you can incorporate the ones that are interesting into your framework.

Other uses I can see of this `Type` is to be used as a message in some errors through `command.Type.FullName`. But this information could be extracted from the command schema itself.

Reply:
Now, I have a much clearer vision.